### PR TITLE
[top_earlgrey] Integrate pwrmgr

### PR DESCRIPTION
**Pull request by tjaychen**
Integrate pwrmgr into top_earlgrey.

Various related fixes were required as part of this process.

### CODE REVIEW

**Comment by tjaychen** [comment]
Would you guys mind having a look at the 3 commits?

The first commit i'll break it out into its own PR, i just wanted to get some review cycles in.

The other two will probably end up as a single PR.

**Comment by sjgitty** [comment]
will try and review tomorrow

File: hw/top_earlgrey/data/xbar_peri.hjson [hunk 1]
```diff
@@ -41,9 +41,15 @@
       clock:     "clk_peri_i",
       reset:     "rst_peri_ni",
       pipeline: "false"
+    },
```
Comment (sriyerg) [inline hw/top_earlgrey/data/xbar_peri.hjson:44]: @weicaiyang

Comment (weicaiyang) [reply hw/top_earlgrey/data/xbar_peri.hjson:44]: xbar TB is automatically updated for this change. LGTM

**Comment by sriyerg** [comment]
LGTM as well.

File: hw/ip/pwrmgr/data/pwrmgr.hjson [hunk 2]
```diff
@@ -70,6 +70,30 @@
...
+      WFI has occurred.
+
+      It unlocks whenever a low power transition has completed (transition back to the
+      ACTIVE state) for any reason.
+      ''',
+
+      fields: [
+        { bits: "0",
+          name: "EN",
+          desc: "Configuration enable.  Once set to 0, it can longer be configured to 1",
```
Comment (msfschaffner) [inline hw/ip/pwrmgr/data/pwrmgr.hjson:90]: nit: it can **no** longer be configured to 1

**Comment by tjaychen** [comment]
ah! thank you.

**Comment by tjaychen** [comment]
any other thoughts on this one?

**Comment by sjgitty** [comment]
From what I can follow, this makes sense. I will admit the intricacies can be hard to follow. We might have to do a cleanup pass in the future, or maybe I just need to stare at it longer.

@tjaychen is this ready to merge now?

File: util/topgen/intermodule.py [hunk 3]
```diff
@@ -530,6 +530,9 @@ def im_netname(obj: OrderedDict, suffix: str = "") -> str:
         if obj["act"] == "rsp" and suffix == "req":
             return "{package}::{struct}_REQ_DEFAULT".format(
                 package=obj["package"], struct=obj["struct"].upper())
+        if obj["act"] == "rcv" and suffix == "":
```
Comment (tjaychen) [inline util/topgen/intermodule.py:533]: @eunchan 

let me know if you're okay with this change. I had to modify this because the `uni` `rcv` types were getting a default value without this change.

**Comment by tjaychen** [comment]
> From what I can follow, this makes sense. I will admit the intricacies can be hard to follow. We might have to do a cleanup pass in the future, or maybe I just need to stare at it longer.
> 
> @tjaychen is this ready to merge now?

sorry @sjgitty i forgot to respond. The answer is not quite because @eunchan updated the intermodule functions, and i wanted to absorb that correctly before merging.  We are almost there. I may also gate this one on #2011

File: util/topgen/intermodule.py (hunk 3 above)
Comment (eunchan) [reply util/topgen/intermodule.py:533]: ... Thanks for catching it!

**Comment by tjaychen** [comment]
i think this is ready to merge now. I'll wait until tomorrow morning to see if there are objections.

File: hw/ip/pwrmgr/data/pwrmgr.hjson [hunk 4]
```diff
@@ -70,6 +77,30 @@
   regwidth: "32",
   registers: [
 
+    { name: "CTRL_CFG_REGWEN",
+      swaccess: "ro",
```
Comment (weicaiyang) [inline hw/ip/pwrmgr/data/pwrmgr.hjson:81]: @tjaychen sorry for the late comment
Is this reg "RO" for SW? It's a bit weird for write enable register.
Is the description "Once set to 0, it can no longer be configured to 1" true for field "EN"? HW will only set it to 0.

@cindychip for visibility

**Comment by tjaychen** [comment]
Thanks weicai, let me clarify that comment.

This one is RO to software because it's a hwext.. so all the conditions are
managed by hardware.

**Comment by weicaiyang** [comment]
> Thanks weicai, let me clarify that comment. This one is RO to software because it's a hwext.. so all the conditions are managed by hardware.

Thanks Tim for confirming.
